frontend/pre_analysis.py

```diff
@@ -0,0 +1,133 @@
+        return 1 if len(func_defs) == 0 else max([len(node.args.args) for node in func_defs])
+
+    def maximum_tuple_length(self):
+        """Get the maximum length of tuples appearing in the AST"""
+        tuples = [node for node in self.all_nodes if isinstance(node, ast.Tuple)]
+        return 0 if len(tuples) == 0 else max([len(node.elts) for node in tuples])
+
+    def classes_pre_analysis(self):
+        # TODO propagate attributes to subclasses.
+        class_defs = [node for node in self.all_nodes if isinstance(node, ast.ClassDef)]
+
+        propagate_attributes_to_subclasses(class_defs)
+
+        class_to_attributes = {}
+        class_to_base = {}
+
+        for cls in class_defs:
+            if len(cls.bases) > 1:
+                raise NotImplementedError("Multiple inheritance is not supported yet.")
+            elif len(cls.bases) == 1:
+                class_to_base[cls.name] = cls.bases[0].id
+            else:
+                class_to_base[cls.name] = "object"
+
+            add_init_if_not_existing(cls)
+
+            attributes = set()
+            class_to_attributes[cls.name] = attributes
+
+            # Inspect all class-level statements
+            for cls_stmt in cls.body:
+                if isinstance(cls_stmt, ast.FunctionDef):
+                    # Add function to class attributes and get attributes defined by self.some_attribute = value
+                    attributes.add(cls_stmt.name)
+                    if len(cls_stmt.args.args) == 0:
+                        continue
+                    first_arg = cls_stmt.args.args[0].arg  # In most cases it will be 'self'
+
+                    # Get attribute assignments where attribute value is the same as the first argument
+                    func_nodes = ast.walk(cls_stmt)
+                    assignments = [node for node in func_nodes if isinstance(node, ast.Assign)]
+
+                    for assignment in assignments:
+                        for target in assignment.targets:
+                            if (isinstance(target, ast.Attribute) and isinstance(target.value, ast.Name) and
+                                    target.value.id == first_arg):
+                                attributes.add(target.attr)
+                elif isinstance(cls_stmt, ast.Assign):
+                    # Get attributes defined as class-level assignment
+                    for target in cls_stmt.targets:
+                        if isinstance(target, ast.Name):
+                            attributes.add(target.id)
+
+        return class_to_attributes, class_to_base
+
+
+def propagate_attributes_to_subclasses(class_defs):
+    inheritance_forest = get_inheritance_forest(class_defs)
+    roots = get_forest_roots(inheritance_forest)
+    name_to_node = class_name_to_node(class_defs)
+
+    for root in roots:
+        propagate(root, inheritance_forest, name_to_node)
+
+
+def propagate(node, inheritance_forest, name_to_node):
```

Comment:
If you have a function with a name like "propagate", which by itself doesn't tell you what the function does, there should be a docstring. Of course, it wouldn't hurt to have more docstrings in general.